### ISSUE TITLE
Ensemble model cannot be inferenced by clients without clear error log to debug.

### ISSUE THREAD

**Issue by Edwardmark** [triton-inference-server/dali_backend#70]
**Description**
I run a ensemble model contains three model which executed sequencely one by one, I check each model, and each one is ok, and I also check two models ensemble, that is ok too. But when I connect them together,and I run the grpc client, the server crashed without meaningful error logs as follows:
```
Traceback (most recent call last):
  File "grpc_client.py", line 209, in <module>
    main()
  File "grpc_client.py", line 182, in main
    results = triton_client.infer(model_name=model_name, inputs=inputs, outputs=outputs)
  File "/usr/local/lib/python3.8/dist-packages/tritonclient/grpc/__init__.py", line 1086, in infer
    raise_error_grpc(rpc_error)
  File "/usr/local/lib/python3.8/dist-packages/tritonclient/grpc/__init__.py", line 61, in raise_error_grpc
    raise get_error_grpc(rpc_error) from None
tritonclient.utils.InferenceServerException: [StatusCode.UNAVAILABLE] Socket closed
``` 

**Triton Information**
21.03 docker container

**To Reproduce**
The ensemble config.pbtxt is as follows:
```
platform: "ensemble"
max_batch_size: 16
input [
  {
    name: "IMAGE_RAW"
    data_type: TYPE_UINT8
    dims: [ -1 ]
  }
]
output [
  {
    name: "SCALE_RATIO" # ratio
    data_type: TYPE_FP32
    dims: [2]
  },
  {
    name: "NUM_DETECTIONS"
    data_type: TYPE_INT32
    dims: [ 1 ]
  },
  {
    name: "NMSED_SCORES"
    data_type: TYPE_FP32
    dims: [ 100 ]
  },
  {
    name: "NMSED_CLASSES"
    data_type: TYPE_FP32
    dims: [ 100 ]
  },
  {
    name: "SCALED_NMSED_BOXES"
    data_type: TYPE_FP32
    dims: [ 100, 4 ]
  }
]

ensemble_scheduling {
  step [
    {
      model_name: "dali_det_pre"
      model_version: -1
      input_map {
        key: "IMAGE_RAW"
        value: "IMAGE_RAW"
      }
      output_map {
        key: "DALI_OUTPUT_0"
        value: "NORM_IMG"
      }
      output_map {
        key: "DALI_OUTPUT_1"
        value: "SCALE_RATIO"
      }
    },
    {
      model_name: "face_det-ucs"
      model_version: -1
      input_map {
        key: "images"
        value: "NORM_IMG"
      }
      output_map {
        key: "num_detections"
        value: "NUM_DETECTIONS"
      }
      output_map {
        key: "nmsed_boxes"
        value: "NMSED_BOXES"
      }
      output_map {
        key: "nmsed_scores"
        value: "NMSED_SCORES"
      }
      output_map {
        key: "nmsed_classes"
        value: "NMSED_CLASSES"
      }
    },
    {
      model_name: "dali_det_post"
      model_version: -1
      input_map {
        key: "NMSED_BOXES"
        value: "NMSED_BOXES"
      }
      input_map {
        key: "SCALE_RATIO_INPUT"
        value: "SCALE_RATIO"
      }
      output_map {
        key: "SCALED_NMSED_BOXES_OUTPUT"
        value: "SCALED_NMSED_BOXES"
      }
    }
  ]
}
```
The client is as follows:
```
FLAGS = parse_args()

    triton_client = tritonclient.grpc.InferenceServerClient(url=FLAGS.url,
                                                            verbose=FLAGS.verbose)

    model_name = FLAGS.model_name
    model_version = -1

    print("Loading images")

    image_data = load_images(FLAGS.img_dir if FLAGS.img_dir is not None else FLAGS.img,
                             max_images=FLAGS.batch_size * FLAGS.n_iter)

    image_data = array_from_list(image_data)
    inputs = generate_inputs(FLAGS.input_names, image_data.shape, "UINT8")
    outputs = generate_outputs(FLAGS.output_names)

    # Initialize the data
    inputs[0].set_data_from_numpy(image_data)
    # Test with outputs
    results = triton_client.infer(model_name=model_name, inputs=inputs, outputs=outputs)
    print(results)
```
**Expected behavior**
results should be obtained without error, but now,  the server just crashes. 
@deadeyegoodwin Looking forward to your reply.

The dali_det_post model config.pbtxt is as follows:
```
backend: "dali"
max_batch_size: 32
input [
  {
    name: "NMSED_BOXES"
    data_type: TYPE_FP32
    dims: [ 100, 4 ]
  },
  {
    name: "SCALE_RATIO_INPUT"
    data_type: TYPE_FP32
    dims: [ 2 ]
  }
]
output [
  {
    name: "SCALED_NMSED_BOXES_OUTPUT" 
    data_type: TYPE_FP32
    dims: [100, 4]
  }
]
dynamic_batching {
  preferred_batch_size: [ 4, 8, 16, 32 ]
  max_queue_delay_microseconds: 100
}
```
the above pipeline is generated using the following code:
```
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import nvidia.dali as dali
import nvidia.dali.fn as fn
import nvidia.dali.types as types

pipe = dali.pipeline.Pipeline(batch_size=32, num_threads=8)
with pipe:
    nmsed_boxes = fn.external_source(device='gpu', name="NMSED_BOXES")
    scale_ratio = fn.external_source(device='gpu', name='SCALE_RATIO_INPUT')
   
    # Rescale BBOX
    ratio = fn.reductions.min(scale_ratio)
    nmsed_boxes /= ratio
    pipe.set_outputs(nmsed_boxes)

pipe.serialize(filename="1/model.dali")
```
The above dali_det_post modle can run correctly itself, but connecting it to the firt two models causes crashes in server.

Change the above post-processing model with python backend model as follows can run without error:
```
for request in requests:
            # Get INPUT0
            in_0 = pb_utils.get_input_tensor_by_name(request, "INPUT0")
            # Get INPUT1
            in_1 = pb_utils.get_input_tensor_by_name(request, "INPUT1")

            out_0 = in_0.as_numpy() / np.min(in_1.as_numpy())

            # Create output tensors. You need pb_utils.Tensor
            # objects to create pb_utils.InferenceResponse.
            out_tensor_0 = pb_utils.Tensor("OUTPUT0",
                                           out_0.astype(output0_dtype))

            # Create InferenceResponse. You can set an error here in case
            # there was a problem with handling this inference request.
            # Below is an example of how you can set errors in inference
            # response:
            #
            # pb_utils.InferenceResponse(
            #    output_tensors=..., TritonError("An error occured"))
            inference_response = pb_utils.InferenceResponse(
                output_tensors=[out_tensor_0])
            responses.append(inference_response)

```
Python config.pbtxt is as follows:
```
name: "python_det_post"
backend: "python"

input [
  {
    name: "INPUT0"
    data_type: TYPE_FP32
    dims: [ 100, 4 ]
    
  }
]
input [
  {
    name: "INPUT1"
    data_type: TYPE_FP32
    dims: [ 2 ]
    
  }
]
output [
  {
    name: "OUTPUT0"
    data_type: TYPE_FP32
    dims: [ 100, 4 ]
  }
]

instance_group [{ kind: KIND_CPU }]

```
Any suggestions please? @deadeyegoodwin Thanks in advance.

**Comment by Edwardmark** [comment]
@szalpal because the network is not always good, so I want to clone repos beforhand, then just use the repo to make the build process more quicklyl.

**Comment by szalpal** [comment]
@Edwardmark ,

I see. It would be possible to tweak the root `CMakeLists.txt` file in order to achieve what you want. Although it is not in our scope right now, so we will not implement it, you would need to try to do it yourself.

**IMPORTANT: this is a dirty explanation of a workaround and we certainly do not support nor plan to support this way of building in foreseeable future. We also highly discourage changing this building procedure for production environments.**

The point is, that there are these three repos, that need to be acquired for proper building any backend: `core`, `common` and `backend`. Our build procedure acquires them in these three declarations:
https://github.com/triton-inference-server/dali_backend/blob/bb9204c5a65734dbb57c3ae9cabadc1e1a870dc3/CMakeLists.txt#L54-L71

Should you like to change them to be acquired from your disk, firstly clone all three repos you need and then you can switch from fetching content from git repository to fetching content from disk location, by changing `GIT`, `GIT_SHALLOW` and `GIT_REPOSITORY` subcommands. Below is the documentation of the `FetchContent` functions, which might be helpful:
https://cmake.org/cmake/help/latest/module/FetchContent.html
https://cmake.org/cmake/help/latest/module/ExternalProject.html#command:externalproject_add
You should pay attention to `Directory Options` in `ExternalProject_Add` directive

**Comment by Edwardmark** [comment]
@szalpal Thank you very much.

**Comment by Edwardmark** [comment]
@szalpal could you please give me more hit on how to change GIT, GIT_SHALLOW and GIT_REPOSITORY subcommands? Thanks.

**Comment by szalpal** [comment]
@Edwardmark ,

what is the problem you are facing?

**Comment by Edwardmark** [comment]
> @szalpal could you please give me more hit on how to change GIT, GIT_SHALLOW and GIT_REPOSITORY subcommands? Thanks. I changed the lines as follows:
> 
> ```
> 
> FetchContent_Declare(
>   repo-common
>   SOURCE_DIR /dali/common/
> )
> FetchContent_Declare(
>   repo-core
>   SOURCE_DIR /dali/core/
> )
> FetchContent_Declare(
>   repo-backend
>   SOURCE_DIR /dali/backend/
> )
> FetchContent_MakeAvailable(repo-common repo-core repo-backend)
> ```
> 
> is that right?
> The DIR /dali/common, /dali/core/, /dali/backend/ is the obtained by :
> 
> ```
>  git clone https://github.com/triton-inference-server/common.git
>  git clone https://github.com/triton-inference-server/core.git
>  git clone https://github.com/triton-inference-server/backend.git
> ```
> 
> I build the docker image successfully.
@szalpal I just want to make sure if the way I tried is correct to replace git repo with local repos.

**Comment by szalpal** [comment]
@Edwardmark ,

As I mentioned above, we do not support nor plan to support this kind of building procedure. Therefore I unfortunately won't be able to answer all the questions, simply because I didn't tried it nor tested it.

The error you're facing is there because the server verifies the API version the backend has been built with. Be sure to use proper version of `backend.git` repo, which has the following defines:
```c++
#define TRITONBACKEND_API_VERSION_MAJOR 1
#define TRITONBACKEND_API_VERSION_MINOR 0
```

**Comment by Edwardmark** [comment]
> @Edwardmark ,
> 
> As I mentioned above, we do not support nor plan to support this kind of building procedure. Therefore I unfortunately won't be able to answer all the questions, simply because I didn't tried it nor tested it.
> 
> The error you're facing is there because the server verifies the API version the backend has been built with. Be sure to use proper version of `backend.git` repo, which has the following defines:
> 
> ```c++
> #define TRITONBACKEND_API_VERSION_MAJOR 1
> #define TRITONBACKEND_API_VERSION_MINOR 0
> ```

So with commit id should I checkout in backend.git to make it the 1.0 vervion? Thanks.

**Comment by Edwardmark** [comment]
@szalpal Do I have to use python3.8? I saw in dockerfile, it install python3.8. Thanks.

**Comment by szalpal** [comment]
Not necessarily. Any DALI version starting from 1.0 should be fine

**Comment by Edwardmark** [comment]
@szalpal Thanks.

**Comment by szalpal** [comment]
> @szalpal Do I have to install nvidia-dali-nightly? 
> https://github.com/triton-inference-server/dali_backend/blob/main/docker/Dockerfile.release#L65
> Thanks.

Not necessarily. We recommend using latest DALI release

**Comment by Edwardmark** [comment]
> https://github.com/triton-inference-server/dali_backend/blob/main/docker/Dockerfile.release#L65

> @Edwardmark
> 
> It's possible, that even though you changed the ExternalSource to `"cpu"`, the bug still prevents normal processing. Anyhow, we've just merged the GPU input feature to upstream. It's going to be released in `tritonserver:21.06`, however it's very easy to run the upstream `dali_backend` with the latest `tritonserver` release.
> 
> Could you try it out and verify if the GPU input solves you problem, or we need to dig deeper? The instructions how to build `dali_backend` docker image are here: [Docker build](https://github.com/triton-inference-server/dali_backend#docker-build-)

If I use dali 1.2,  would the dali_backend support gpu input?

**Comment by szalpal** [comment]
> If I use dali 1.2, would the dali_backend support gpu input?

@Edwardmark, yes. Although we don't guarantee backwards compatibility. Therefore, only the latest DALI version is properly tested and maintained
